cpu/stm32f3/periph/i2c.c

```diff
@@ -0,0 +1,490 @@
+
+    /* wait for ack */
+    DEBUG("Waiting for ACK\n");
+    while (!(i2c->ISR & I2C_ISR_TXIS));
+
+    /* send register number */
+    DEBUG("ACK received, write reg into DR\n");
+    i2c->TXDR = reg;
+
+    /* send repeated start sequence, read registers and end transmission */
+    DEBUG("ACK received, send repeated start sequence\n");
+    return i2c_read_bytes(dev, address, data, length);
+}
+
+int i2c_write_byte(i2c_t dev, uint8_t address, char data)
+{
+    return i2c_write_bytes(dev, address, &data, 1);
+}
+
+int i2c_write_bytes(i2c_t dev, uint8_t address, char *data, int length)
+{
+    I2C_TypeDef *i2c;
+
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            i2c = I2C_0_DEV;
+            break;
+#endif
+
+        default:
+            return -1;
+    }
+
+    /* start transmission and send slave address */
+    _start(i2c, address, length, I2C_FLAG_WRITE);
+
+    /* send out data bytes */
+    _write(i2c, data, length);
+
+    /* end transmission */
+    _stop(i2c);
+
+    return length;
+}
+
+int i2c_write_reg(i2c_t dev, uint8_t address, uint8_t reg, char data)
+{
+    return i2c_write_regs(dev, address, reg, &data, 1);
+}
+
+int i2c_write_regs(i2c_t dev, uint8_t address, uint8_t reg, char *data, int length)
+{
+    I2C_TypeDef *i2c;
+
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            i2c = I2C_0_DEV;
+            break;
+#endif
+
+        default:
+            return -1;
+    }
+
+    /* start transmission and send slave address */
+    /* increase length because our data is register+data */
+    _start(i2c, address, length+1, I2C_FLAG_WRITE);
+
+    /* wait for ack */
+    DEBUG("Waiting for ACK\n");
+    while (!(i2c->ISR & I2C_ISR_TXIS));
+
+    /* send register number */
+    DEBUG("ACK received, write reg into DR\n");
+    i2c->TXDR = reg;
+
+    /* write out data bytes */
+    _write(i2c, data, length);
+
+    /* end transmission */
+    _stop(i2c);
+
+    return length;
+}
+
+void i2c_poweron(i2c_t dev)
+{
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            I2C_0_CLKEN();
+            break;
+#endif
+    }
+}
+
+void i2c_poweroff(i2c_t dev)
+{
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            while (I2C_0_DEV->ISR & I2C_ISR_BUSY);
+
+            I2C_0_CLKDIS();
+            break;
+#endif
+    }
+}
+
+static void _start(I2C_TypeDef *dev, uint8_t address, uint8_t length, uint8_t rw_flag)
+{
+    /* set address mode to 7-bit */
+    dev->CR2 &= ~(I2C_CR2_ADD10);
+
+    /* set slave address */
+    dev->CR2 &= ~(I2C_CR2_SADD);
+    dev->CR2 |= (address << 1);
+
+    /* set transfer direction */
+    dev->CR2 &= ~(I2C_CR2_RD_WRN);
+    dev->CR2 |= (rw_flag << 10);
+
+    /* set number of bytes */
+    dev->CR2 &= ~(I2C_CR2_NBYTES);
+    dev->CR2 |= (length << 16);
+
+    /* configure autoend configuration */
+    dev->CR2 &= ~(I2C_CR2_AUTOEND);
+
+    /* generate start condition */
+    DEBUG("Generate start condition\n");
+    dev->CR2 |= I2C_CR2_START;
+}
```

Comment:
Clock synchronization is implemented in master mode using the timing register settings. See datasheet 28.4.8.